neo/core/analogsignal.py

```diff
@@ -446,34 +446,38 @@ def _pp(line):
 
     def time_index(self, t):
         """Return the array index corresponding to the time `t`"""
-        t = t.rescale(self.sampling_period.units)
-        i = (t - self.t_start) / self.sampling_period
-        i = int(np.rint(i.magnitude))
+        i = (t - self.t_start) * self.sampling_rate
+        i = int(np.rint(i.simplified.magnitude))
         return i
 
     def time_slice(self, t_start, t_stop):
         '''
         Creates a new AnalogSignal corresponding to the time slice of the
         original AnalogSignal between times t_start, t_stop. Note, that for
-        numerical stability reasons if t_start, t_stop do not fall exactly on
-        the time bins defined by the sampling_period they will be rounded to
-        the nearest sampling bins.
+        numerical stability reasons if t_start does not fall exactly on
+        the time bins defined by the sampling_period it will be rounded to
+        the nearest sampling bin. The time bin for t_stop will be chosen to
+        make the duration of the resultant signal as close as possible to
+        t_stop - t_start. This means that for a given duration, the size
+        of the slice will always be the same.
         '''
 
         # checking start time and transforming to start index
         if t_start is None:
             i = 0
+            t_start = 0 * pq.s
         else:
             i = self.time_index(t_start)
 
         # checking stop time and transforming to stop index
         if t_stop is None:
             j = len(self)
         else:
-            j = self.time_index(t_stop)
+            duration = (t_stop - t_start).rescale(self.sampling_period.units)
+            j = i + int(np.rint(duration * self.sampling_rate))
```

Comment:
Maybe here you can also use simplified, no ?
```j = i + int(np.rint((duration * self.sampling_rate)).simplified)```

Reply:
good idea. I've pushed a new commit. I think this can be merged now.